core/src/main/java/com/yahoo/oak/Chunk.java

```diff
@@ -320,27 +334,29 @@ private int binaryFind(KeyBuffer tempKey, K key) {
 
         // if the first item is already larger than key,
         // return NONE_NEXT to indicate that a regular linear search is needed
-        if (compareKeyAndEntryIndex(tempKey, key, entrySet.getHeadNextIndex()) <= 0) {
+        if (compareKeyAndEntryIndex(tempKey, key, entrySet.getHeadNextEntryIndex()) <= 0) {
             return NONE_NEXT;
         }
 
         // optimization: compare with last key to avoid binary search (here sortedCount is not zero)
-        if (compareKeyAndEntryIndex(tempKey, key, sortedCount) > 0) {
-            return sortedCount;
+        if (compareKeyAndEntryIndex(tempKey, key, sortedCount - 1) > 0) {
+            return sortedCount - 1;
         }
 
-        int start = 0;
+        // `start` and `end` are intentionally initiated outside of the array boundaries.
```

Comment:
So the returned key as exactly the key we are looking for, or the maximal one still less than the key we are looking for...

Reply:
fixed